Benchmark dataclass with `slots=True`. Benchmark struct access (not creation), add `uv` command to README

polyrand opened the pull request:
Hi! Thanks for these benchmarks. I'm making this "draft" PR mainly as a point of discussion, I put different (unrelated) changes which could probably be split into different PRs if any of these changes are welcome.

1. Benchmark dataclasses with `slots=True`

This is my first go-to when trying to make programs more memory efficient. I know this benchmark doesn't measure memory usage, but I think it could be a nice "good to have". `@dataclass(slots=True)` is only available in since Python 3.10, thought. If compatibility with python >= 3.8 is needed, I can add the `__slots__` manually.

2. Benchmark attribute access

I think one issue with the "composite-structs" benchmarks is that it mixes object creation with attribute access. But in some cases, one of those actions may be more prevalent than the other. I add a `composite-structs-access` group to measure attribute access only. It shows different results than then ones in `composite-structs` (although, surprisingly, the `namedtuple` access is still the slowest)

3. Better reproducibility

~I noticed that the `pyproject.toml` file doesn't properly list all the dependencies needed to run the script~

I see this was just fixed in https://github.com/ashvardanian/less_slow.py/commit/9a5ffbb86419e64fa8c6ae615e415bb57fba374e

In any case, I found it useful to have a single `uv` command that allows running the script, and could even be useful to run the benchmark with different Python versions. I'm not sure if this is something that could be nice to add to the README.

polyrand committed:
Benchmark dataclass with slots

polyrand committed:
Add benchmarks (and comment) for struct access

polyrand committed:
Add uv command to README

ashvardanian commented:
Great additions! Will merge later today 🤗

polyrand commented:
Thanks!

If we want to keep `@dataclass(slots=True)`, the [`requires-python` value of pyproject.toml](https://github.com/ashvardanian/less_slow.py/blob/fc9516cf67d94fa4ce6f5dec06715751b3fa5245/pyproject.toml#L9) would have to be updated to say `>=3.10` for correctness.

Otherwise, I can add this change. Whatever you think is better:

```diff
-@dataclass(slots=True)
+@dataclass
 class PointSlotsDataclass:
+    __slots__ = ('x', 'y', 'flag')
     x: float
     y: float
     flag: bool
```

polyrand commented:
Another interesting finding is that the order changes depending on the Python version. On my laptop (MacBook M1 Pro, 32GB RAM), and filtering the tests by `test_struct`:

## Python 3.12

```sh
uv run --python-preference="only-managed" --python="3.12" --no-sync --with "pytest" --with "pytest-benchmark" --with "numpy" --with "pandas" --with "pyarrow" pytest -ra -q -k test_struct --benchmark-columns="min, max, mean, stddev, median, outliers, rounds, iterations" less_slow.py
```

Results:

```
-------------------------------------------------------------- benchmark 'composite-structs': 7 tests --------------------------------------------------------------
Name (time in ns)                     Min                    Max                Mean              StdDev              Median            Outliers  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_structs_tuple_unpacking      43.9187 (1.0)         727.8559 (1.53)      45.3137 (1.0)        4.2477 (1.30)      45.0451 (1.0)      232;7436  200000         111
test_structs_tuple_indexing       45.5606 (1.04)        474.6916 (1.0)       46.7669 (1.03)       3.2740 (1.0)       46.7288 (1.04)     380;8897  200000         107
test_structs_dict                 87.9150 (2.00)        655.2051 (1.38)      89.4580 (1.97)       5.2310 (1.60)      89.1650 (1.98)     306;4157   55556         200
test_structs_dataclass           124.9718 (2.85)     20,208.0118 (42.57)    189.1595 (4.17)     113.5881 (34.69)    207.9760 (4.62)      350;409  186047           1
test_structs_namedtuple          125.0010 (2.85)     16,292.0041 (34.32)    244.3309 (5.39)     104.3919 (31.88)    250.0019 (5.55)    363;39960  167842           1
test_structs_slots_dataclass     128.3298 (2.92)      1,157.9102 (2.44)     131.2327 (2.90)       9.2306 (2.82)     130.4198 (2.90)     301;9998   75712         100
test_structs_class               137.4996 (3.13)      3,427.0997 (7.22)     145.3260 (3.21)      30.8698 (9.43)     143.7511 (3.19)     188;5143  166668          20
--------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------- benchmark 'composite-structs-access': 7 tests ------------------------------------------------------------
Name (time in ns)                           Min                   Max               Mean            StdDev             Median             Outliers  Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_structs_tuple_unpacking_access     39.6340 (1.0)        594.5122 (1.08)     40.8123 (1.0)      4.0784 (1.08)     40.6505 (1.0)      356;81685  198373         123
test_structs_slots_dataclass_access     42.3914 (1.07)       585.1391 (1.06)     43.8826 (1.08)     5.3897 (1.43)     43.4783 (1.07)     619;21724  198374         115
test_structs_class_access               43.1414 (1.09)       599.9294 (1.09)     44.5115 (1.09)     3.7696 (1.0)      44.2479 (1.09)      437;9399  198374         113
test_structs_dataclass_access           43.1414 (1.09)       564.1592 (1.02)     44.5024 (1.09)     3.9579 (1.05)     44.2479 (1.09)      230;7140  198334         113
test_structs_tuple_indexing_access      45.0001 (1.14)       552.0799 (1.0)      46.4700 (1.14)     4.2385 (1.12)     46.2501 (1.14)     232;96458  200000         100
test_structs_dict_access                48.7501 (1.23)     1,067.5000 (1.93)     50.3101 (1.23)     5.3140 (1.41)     50.0001 (1.23)     328;10140  196697         100
test_structs_namedtuple_access          57.4999 (1.45)       716.2499 (1.30)     58.8328 (1.44)     5.8573 (1.55)     58.7499 (1.45)     324;11481  165509         100
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
```


## Python 3.9

```sh
uv run --python-preference="only-managed" --python="3.9" --no-sync --with "pytest" --with "pytest-benchmark" --with "numpy" --with "pandas" --with "pyarrow" pytest -ra -q -k test_struct --benchmark-columns="min, max, mean, stddev, median, outliers, rounds, iterations" less_slow.py
```

Results:

```
-------------------------------------------------------------- benchmark 'composite-structs': 7 tests --------------------------------------------------------------
Name (time in ns)                     Min                    Max                Mean              StdDev              Median            Outliers  Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_structs_tuple_unpacking      57.0800 (1.0)         752.0900 (1.0)       58.3151 (1.0)        4.7243 (1.0)       57.9200 (1.0)      265;7668  164393         100
test_structs_tuple_indexing       63.7500 (1.12)        810.0000 (1.08)      65.6000 (1.12)       7.4193 (1.57)      65.0000 (1.12)    395;24030  136352         100
test_structs_dict                101.4550 (1.78)        881.6650 (1.17)     102.8573 (1.76)       8.9428 (1.89)     102.0850 (1.76)     428;7433   47430         200
test_structs_slots_dataclass     125.0000 (2.19)     16,167.0000 (21.50)    228.2549 (3.91)      80.5139 (17.04)    209.0000 (3.61)      428;428  192013           1
test_structs_class               166.0000 (2.91)     18,250.0000 (24.27)    253.7357 (4.35)     102.2062 (21.63)    250.0000 (4.32)    380;16038  145455           1
test_structs_namedtuple          192.0800 (3.37)      1,111.2500 (1.48)     197.4507 (3.39)      10.7485 (2.28)     196.6600 (3.40)     647;5712   48881         100
test_structs_dataclass           208.3043 (3.65)      2,742.7826 (3.65)     215.8330 (3.70)      17.0263 (3.60)     215.5652 (3.72)     200;8370  196735          23
--------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------- benchmark 'composite-structs-access': 7 tests ------------------------------------------------------------
Name (time in ns)                           Min                   Max               Mean             StdDev             Median            Outliers  Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_structs_tuple_unpacking_access     50.8300 (1.0)      8,506.2500 (13.82)    53.4395 (1.0)      55.3394 (11.28)    52.0800 (1.0)      712;8027  179084         100
test_structs_tuple_indexing_access      63.3300 (1.25)       790.8300 (1.29)     64.6516 (1.21)      5.6720 (1.16)     64.5800 (1.24)     251;7387  151884         100
test_structs_dict_access                66.6600 (1.31)     1,125.4200 (1.83)     68.1649 (1.28)      6.4060 (1.31)     67.9200 (1.30)    268;40298  146349         100
test_structs_namedtuple_access          67.5000 (1.33)     1,997.5000 (3.25)     69.9921 (1.31)     11.6309 (2.37)     68.7500 (1.32)     830;2382  143720         100
test_structs_slots_dataclass_access     69.5800 (1.37)       615.4200 (1.0)      71.0979 (1.33)      4.9045 (1.0)      70.8300 (1.36)     240;8900  142006         100
test_structs_class_access               71.2500 (1.40)       690.0000 (1.12)     73.1929 (1.37)      6.5219 (1.33)     72.5000 (1.39)    494;10356  136370         100
test_structs_dataclass_access           71.2500 (1.40)       797.0900 (1.30)     73.4756 (1.37)      6.2922 (1.28)     72.9100 (1.40)    525;18209  138735         100
----------------------------------------------------------------------------------------------------------------------------------------------------------------------

```


This may be an important caveat to mention in the README. The "faster" method may be different across Python versions.

ashvardanian commented:
I am familiar with this artifact, the behavior of different snippets very much depends on the Python version.

As for dataclasses, we can have 2 variants, one with slots and one without - conditionally evaluated.

polyrand committed:
Keep code compatible with Python <3.10

The `slots` parameter was added to dataclasses in Python 3.10. This commit changes the code to explicitly set the `__slots__` attribute instead of realying on `@dataclass(slots=True)`

The benchmark results are not affected by the change. The `test_structs_slots_dataclass_access` benchmark still reports a mean time of 43.8826ns

polyrand commented:
> we can have 2 variants, one with slots and one without - conditionally evaluated.

I'm not sure if I understand what that means. We can benchmark `__slots__` in both cases. The question was more about whether to use the `__slots__` syntax, or the `slots=True` parameter from the `dataclass` decorator, which is mainly syntactic sugar.

I'll use the explicit `__slots__` for now to keep backwards compatibility.

polyrand committed:
Rename functions for easier filtering

Rename functions from `test_structs_*_access` to `test_structs_access_*`

This makes it easier to filter tests with `pytest ... -k test_struct_access ...`

ashvardanian commented:
Let's merge and continue incrementally 🤗

ashvardanian commented:
I don't think much experience with `uv`. @polyrand, do you think it may be a good choice for [heterogeneous computing pipelines](https://github.com/ashvardanian/cpp-cuda-python-starter-kit) as well?

ashvardanian commented on less_slow.py:
```diff
@@ -597,6 +615,101 @@ def kernel():
+    assert result == 3.0
+
+
+# ? When accessing pre-constructed objects, the performance gap narrows
```
I am afraid this benchmarking tool isn’t accurate to a nano-second resolution. It’s wiser to repeat more member accesses inside the kernel function for higher precision.

polyrand replied:
Yes, I think I'll keep experimenting with better ways to benchmark the attribute access to get more reliable timings, then open a separate PR just for that.

ashvardanian commented on less_slow.py:
```diff
@@ -597,6 +615,101 @@ def kernel():
+    x, y, _ = (1.0, 2.0, True)
+
+    def kernel():
+        return x + y
```
This kernel performs no member lookup, just addition.

ashvardanian commented on less_slow.py:
```diff
@@ -553,6 +553,24 @@ def kernel():
     assert result == 3.0
 
 
+@dataclass
```
It may be a wise idea to merge slots and uv first and reconsider the rest in a separate PR 🤗

polyrand replied:
Agree! I'll open a separate PR tomorrow with just those two changes.

polyrand commented:
> do you think it may be a good choice for [heterogeneous computing pipelines](https://github.com/ashvardanian/cpp-cuda-python-starter-kit) as well?

I'd need to look into it. I'm not familiar with using `uv` with scripts that require external compilation (setuptools + build_ext).

polyrand commented:
I've now created #2 

This PR can be closed when/if needed.
